src/verify.ts

```diff
@@ -13,106 +13,191 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 import { crypto, dsse, pem } from './util';
 
-export type GetPublicKeyFunc = (keyId: string) => Promise<string | undefined>;
-
-export interface VerifyOptions {
-  tlog: TLog;
-  tlogKeys: Record<string, KeyObject>;
-  getPublicKey?: GetPublicKeyFunc;
-}
+export type KeySelector = (
+  hint: string,
+  trustedKeys: sigstore.PublicKey[]
+) => sigstore.PublicKey | undefined;
 
 export class Verifier {
-  private tlog: TLog;
-  private tlogKeys: Record<string, KeyObject>;
-  private getExternalPublicKey: GetPublicKeyFunc;
-
-  constructor(options: VerifyOptions) {
-    this.tlog = options.tlog;
-    this.tlogKeys = options.tlogKeys;
-    this.getExternalPublicKey =
-      options.getPublicKey || (() => Promise.resolve(undefined));
+  private trustedRoot: sigstore.TrustedRoot;
+  private keySelector: KeySelector;
+
+  constructor(trustedRoot: sigstore.TrustedRoot, keySelector?: KeySelector) {
+    this.trustedRoot = trustedRoot;
+    this.keySelector = keySelector || ((_, keys) => keys[0]);
+  }
+
+  // Verifies the bundle signature, the bundle's certificate chain (if present)
+  // and the bundle's transparency log entries.
+  public verify(
+    bundle: sigstore.Bundle,
+    options: sigstore.RequiredArtifactVerificationOptions,
+    data?: Buffer
+  ): void {
+    this.verifyArtifactSignature(bundle, options, data);
+
+    if (sigstore.isBundleWithCertificateChain(bundle)) {
+      this.verifySigningCertificate(bundle, options);
+    }
+
+    this.verifyTLogEntries(bundle, options);
+  }
+
+  // Performs bundle signature verification. Determines the type of the bundle
+  // content and delegates to the appropriate signature verification function.
+  private verifyArtifactSignature(
+    bundle: sigstore.Bundle,
+    options: sigstore.ArtifactVerificationOptions,
+    data?: Buffer
+  ): void {
+    const publicKey = this.getPublicKey(bundle, options);
+
+    switch (bundle.content?.$case) {
+      case 'messageSignature':
+        if (!data) {
+          throw new VerificationError(
+            'no data provided for message signature verification'
+          );
+        }
+        verifyMessageSignature(
+          data,
+          bundle.content.messageSignature,
+          publicKey
+        );
+        break;
+      case 'dsseEnvelope':
+        verifyDSSESignature(bundle.content.dsseEnvelope, publicKey);
+        break;
+      default:
+        throw new InvalidBundleError('no content found');
+    }
   }
 
-  public async verifyOffline(bundle: Bundle, data?: Buffer): Promise<void> {
-    const publicKey = await this.getPublicKey(bundle);
+  // Performs verification of the bundle's certificate chain. The bundle must
+  // contain a certificate chain and the options must contain the required
+  // options for CA verification.
+  // TODO: We've temporarily removed the requirement that the options contain
+  // the list of trusted signer identities. This will be added back in a future
+  // release.
+  private verifySigningCertificate(
```

Comment:
I changed the way that I'm importing this function (as well as verifyTLogEntries) so that when we invoke them here we can do:
```typescript
ca.verifySigningCertificate(...)
tlog.verifyTLogEntries(...)
```

Reply:
nice one!